torch/_functorch/autograd_function.py

```diff
@@ -686,21 +686,35 @@ class AutogradFunctionApply(HigherOrderOperator):
     def __init__(self):
         super().__init__("autograd_function_apply")
 
-    def __call__(self, fwd, bwd, *fwd_args):
+    def __call__(self, fwd, bwd, *fwd_args, **fwd_kwargs):
         saved_values = None
+        args_tensor_mask = fwd_kwargs["args_tensor_mask"]
+        length_of_tensor_args = sum(args_tensor_mask)
+        # Filter out the original tensor args from fwd_args,
+        # lifted freevars should not be args of ApplyTemplate.apply
+        # since we don't need to calculate the gradients of them.
+        new_fwd_args = fwd_args[:length_of_tensor_args]
 
         class ApplyTemplate(torch.autograd.Function):
             @staticmethod
             def forward(ctx, *args):
                 nonlocal saved_values
-                output, saved_values = fwd(None, *args)
+                output, saved_values = fwd(None, *fwd_args)
                 return output
 
             @staticmethod
             def backward(ctx, *grad):
-                return bwd(None, *grad, *saved_values)
-
-        return ApplyTemplate.apply(*fwd_args)
+                output = bwd(None, *grad, *saved_values)
+                # The output includes the gradients of all original args, which can be None for non-Tensor args.
+                # We need to filter out the None values.
+                if isinstance(output, tuple):
+                    return tuple(
+                        [x for x, mask in zip(output, args_tensor_mask) if mask]
+                    )
+                else:
+                    return output
```

Comment:
Btw, @Chillee's suggestion at the sync might have been to do the rewrite of the backward graph in Dynamo:
1) speculate subgraph tells us which arguments we actually used. In the weird case, it would say something like `(from_arg(0), from_arg(2), "freevar")`
2) Based on this, we would get Dynamo to rewrite the `bwd` graph to drop output 1:
```
# rewrite the following
def bwd(ctx, grad, saved_vars):
    return <Tensor>, None, <Tensor>

# into this:
def bwd(ctx, grad, saved_vars):
    return <Tensor>, <Tensor>
```
All of this would happen in Dynamo (and outside of the autograd_function HOP (this file))

Reply:
Oh, I think we just need to move the rewrite from this file to _dynamo/variables/higher_order_ops.py, which I think it's doable. I'm curious why that option is better than the current one?

Reply:
tbh it's probably easier to do the backward rewrite in this file. I think Horace's point is more of (1): speculate subgraph should tell us which arguments we actually used (this is useful for other HOPs)